oxidation/libparsec/crates/platform-async/src/native/task.rs

```diff
@@ -0,0 +1,166 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BSLv1.1 (eventually AGPLv3) 2016-2021 Scille SAS
+
+use futures::FutureExt;
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context, Poll, Waker},
+};
+
+/// Spawns a new asynchronous task, returning a [Task] for it.
+///
+/// ```
+/// use libparsec_platform_async::{spawn, Timer};
+/// use std::time::Duration;
+///
+/// # #[tokio::main]
+/// # async fn main() {
+/// let task = spawn(async {
+///     Timer::after(Duration::from_millis(20));
+///     42
+/// });
+/// let result = task.await;
+/// assert_eq!(result, 42);
+/// # }
+/// ```
+pub fn spawn<F>(future: F) -> Task<F::Output>
+where
+    F: Future + Send + 'static,
+    F::Output: Send + 'static,
+{
+    let shared_state = SharedState::default();
+    let shared_state = Arc::new(Mutex::new(shared_state));
+
+    let task = Task::new(shared_state.clone());
+    let runnable = Runnable::new(future, shared_state);
+
+    tokio::spawn(runnable);
+
+    task
+}
+
+struct SharedState<T> {
+    canceled: bool,
+    finished: bool,
+    detached: bool,
+    value: Option<T>,
+    task_waker: Option<Waker>,
+    runnable_waker: Option<Waker>,
+}
+
+impl<T> Default for SharedState<T> {
+    fn default() -> Self {
+        Self {
+            canceled: false,
+            finished: false,
+            detached: false,
+            value: None,
+            task_waker: None,
+            runnable_waker: None,
+        }
+    }
+}
+
+pub struct Task<T> {
+    shared_state: Arc<Mutex<SharedState<T>>>,
+}
+
+impl<T> Task<T> {
+    fn new(shared_state: Arc<Mutex<SharedState<T>>>) -> Self {
+        Self { shared_state }
+    }
+
+    /// Cancels the task
+    pub fn cancel(&self) -> Option<T> {
```

Comment:
cancel is too close to what we use in trio, we should use another more explicit name (to show the fact the future is not longer going to be called anymore)

so maybe `abort` to use the tokio jargon